File: pkg/collector/corechecks/servicediscovery/usm/python.go
// Unless explicitly stated otherwise all files in this repository are licensed
// under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2016-present Datadog, Inc.

package usm

import (
	"io/fs"
	"path"
	"path/filepath"
	"strings"
)

const (
	initPy             = "__init__.py"
	allPyFiles         = "*.py"
	gunicornEnvCmdArgs = "GUNICORN_CMD_ARGS"
	wsgiAppEnv         = "WSGI_APP"
)

type pythonDetector struct {
	ctx      DetectionContext
	gunicorn detector
}

type gunicornDetector struct {
	ctx DetectionContext
}

func newGunicornDetector(ctx DetectionContext) detector {
	return &gunicornDetector{ctx: ctx}
}

func newPythonDetector(ctx DetectionContext) detector {
	return &pythonDetector{ctx: ctx, gunicorn: newGunicornDetector(ctx)}
}

type argType int

const (
	argNone argType = iota
	argMod
	argFileName
)

// parsePythonArgs parses the CPython command line arguments to find the module
// name or the file name. For this, besides handling the -m option to get the
// module name, we need to specifically handle any option that is known to take
// an argument so that we can skip the argument and not misinterpret it as a
// filename.
//
// We assume that all the other options other than the ones explicitly handled
// below do not take any arguments.
func parsePythonArgs(args []string) (argType, string) {
	skipNext := false
	modNext := false
	for _, arg := range args {
		if modNext {
			return argMod, arg
		}

		if skipNext {
			skipNext = false
			continue
		}

		if strings.HasPrefix(arg, "--") {
			// Only long arg with an argument.  CPython doesn't allow
			// including the argument with an equals sign in the same arg.
			if arg == "--check-hash-based-pycs" {
				skipNext = true
			}
		} else if strings.HasPrefix(arg, "-") {
		INNER:
			for charidx, char := range arg[1:] {
				rest := arg[1+charidx+1:]
				switch char {
				case 'c':
					// Everything after -c is a command and it terminates
					// the option parsing.
					return argNone, ""
				case 'm':
					// Module name, either attached here or in the next arg.
					if len(rest) > 0 {
						return argMod, rest
					}

					modNext = true
				case 'X', 'W':
					// Takes an argument, either attached here or in the next arg.
					if len(rest) > 0 {
						break INNER
					}

					skipNext = true
				}
			}
		} else {
			return argFileName, arg
		}
	}
	return argNone, ""
}

func (p pythonDetector) detect(args []string) (ServiceMetadata, bool) {
	// When Gunicorn is invoked via its wrapper script the command line ends up
	// looking like the example below, so redirect to the Gunicorn detector for
	// this case:
	//  /usr/bin/python3 /usr/bin/gunicorn foo:app()
	//
	// Another case where we want to redirect to the Gunicorn detector is when
	// gunicorn replaces its command line with something like the below. Because
	// of the [ready], we end up here first instead of going directly to the
	// Gunicorn detector.
	//  [ready] gunicorn: worker [airflow-webserver]
	if len(args) > 0 {
		base := filepath.Base(args[0])
		if base == "gunicorn" || base == "gunicorn:" {
			return p.gunicorn.detect(args[1:])
		}
		if base == "uvicorn" {
			return detectUvicorn(args[1:])
		}
	}

	argType, arg := parsePythonArgs(args)
	switch argType {
	case argNone:
		return ServiceMetadata{}, false
	case argMod:
		return NewServiceMetadata(arg, CommandLine), true
	case argFileName:
		absPath := p.ctx.resolveWorkingDirRelativePath(arg)
		fi, err := fs.Stat(p.ctx.fs, absPath)
		if err != nil {
			return ServiceMetadata{}, false
		}
		stripped := absPath
		var filename string
		if !fi.IsDir() {
			stripped, filename = path.Split(stripped)
			// If the path is a root level file, return the filename
			if stripped == "" {
				return NewServiceMetadata(p.findNearestTopLevel(filename), CommandLine), true
			}
		}
		if value, ok := p.deducePackageName(path.Clean(stripped), filename); ok {
			return NewServiceMetadata(value, Python), true
		}

		name := p.findNearestTopLevel(stripped)
		// If we have generic/useless directory names, fallback to the filename.
		if name == "." || name == "/" || name == "bin" || name == "sbin" {
			name = p.findNearestTopLevel(filename)
		}

		return NewServiceMetadata(name, CommandLine), true
	}

	return ServiceMetadata{}, false
}

// deducePackageName is walking until a `__init__.py` is not found. All the dir traversed are joined then with `.`
func (p pythonDetector) deducePackageName(fp string, fn string) (string, bool) {
	up := path.Dir(fp)
	current := fp
	var traversed []string
	for run := true; run; run = current != up {
		if _, err := fs.Stat(p.ctx.fs, path.Join(current, initPy)); err != nil {
			break
		}
		traversed = append([]string{path.Base(current)}, traversed...)
		current = up
		up = path.Dir(current)
	}
	if len(traversed) > 0 && len(fn) > 0 {
		traversed = append(traversed, strings.TrimSuffix(fn, path.Ext(fn)))
	}
	return strings.Join(traversed, "."), len(traversed) > 0

}

// findNearestTopLevel returns the top level dir the contains a .py file starting walking up from fp.
// If fp is a file, it returns the filename without the extension.
func (p pythonDetector) findNearestTopLevel(fp string) string {
	up := path.Dir(fp)
	current := fp
	last := current
	for run := true; run; run = current != up {
		if matches, err := fs.Glob(p.ctx.fs, path.Join(current, allPyFiles)); err != nil || len(matches) == 0 {
			break
		}
		last = current
		current = up
		up = path.Dir(current)
	}
	filename := path.Base(last)
	return strings.TrimSuffix(filename, path.Ext(filename))
}

func (g gunicornDetector) detect(args []string) (ServiceMetadata, bool) {
	if fromEnv, ok := extractEnvVar(g.ctx.Envs, gunicornEnvCmdArgs); ok {
		name, ok := extractGunicornNameFrom(strings.Split(fromEnv, " "))
		if ok {
			return NewServiceMetadata(name, Gunicorn), true
		}
	}
	if wsgiApp, ok := extractEnvVar(g.ctx.Envs, wsgiAppEnv); ok && len(wsgiApp) > 0 {
		return NewServiceMetadata(parseNameFromWsgiApp(wsgiApp), Gunicorn), true
	}

	if name, ok := extractGunicornNameFrom(args); ok {
		return NewServiceMetadata(name, CommandLine), true
	}
	return NewServiceMetadata("gunicorn", CommandLine), true
}

func extractGunicornNameFrom(args []string) (string, bool) {
	if len(args) == 0 {
		return "", false
	}

	// gunicorn sometimes replaces the cmdline with something like "gunicorn: master [package]".
	lastArg := args[len(args)-1]
	if len(lastArg) >= 2 && lastArg[0] == '[' && lastArg[len(lastArg)-1] == ']' {
		// Extract the name between the brackets
		name := lastArg[1 : len(lastArg)-1]
		return parseNameFromWsgiApp(name), true
	}

	// If the command line is not replaced, we need to parse the arguments.
	// Prefer the --name argument if one exists, otherwise try to find the first
	// non-flag argument.

	// List of long options that do NOT take an argument.  This list is shorter
	// than the ones which do take an argument.
	noArgOptions := map[string]struct{}{
		"--reload":                            {},
		"--spew":                              {},
		"--check-config":                      {},
		"--print-config":                      {},
		"--preload":                           {},
		"--no-sendfile":                       {},
		"--reuse-port":                        {},
		"--daemon":                            {},
		"--initgroups":                        {},
		"--capture-output":                    {},
		"--log-syslog":                        {},
		"--enable-stdio-inheritance":          {},
		"--disable-redirect-access-to-syslog": {},
		"--proxy-protocol":                    {},
		"--suppress-ragged-eofs":              {},
		"--do-handshake-on-connect":           {},
		"--strip-header-spaces":               {},
	}

	nameNext := false
	skipNext := false
	candidate := ""

	for i := 0; i < len(args); i++ {
		arg := args[i]
		if nameNext {
			return arg, true
		}
		if skipNext {
			skipNext = false
			continue
		}

		if strings.HasPrefix(arg, "--name=") {
			// --name=VALUE
			return arg[len("--name="):], true
		} else if arg == "--name" {
			// --name VALUE
			nameNext = true
		} else if strings.HasPrefix(arg, "--") {
			// Skip --option=value
			if strings.Contains(arg, "=") {
				continue
			}
			// Skip the next argument unless we know that the option does not
			// take an argument
			if _, ok := noArgOptions[arg]; !ok {
				skipNext = true
			}
		} else if strings.HasPrefix(arg, "-") {
			// Single letter flags can be grouped together, e.g. "-Rnfoo"
		INNER:
			for charidx, char := range arg[1:] {
				rest := arg[1+charidx+1:]
				switch char {
				case 'n':
					if len(rest) > 0 {
						// Argument attached here
						return rest, true
					}
					// Argument in separate arg
					nameNext = true
				case 'R', 'd':
					// These are the only single letter flags that do not take an argument
					continue
				default:
					// Flag with argument
					if len(rest) > 0 {
						// Argument attached here
						break INNER
					}

					// Argument in separate arg
					skipNext = true
				}
			}
		} else if candidate == "" {
			candidate = arg
		}
	}

	if len(candidate) > 0 {
		// We didn't find a name flag, so try to parse the name from the first
		// potential app/module name.
		return parseNameFromWsgiApp(candidate), true
	}

	return "", false
}

func parseNameFromWsgiApp(wsgiApp string) string {
	name, _, _ := strings.Cut(wsgiApp, ":")
	return name
}

func detectUvicorn(args []string) (ServiceMetadata, bool) {
	skipNext := false
	for _, arg := range args {
		if skipNext {
			skipNext = false
			continue
		}

		// Skip flags
		if strings.HasPrefix(arg, "-") {
			if arg == "--header" {
				// This takes an argument which looks similar to a module:app
				// pattern so skip that.
				skipNext = true
			}
			continue
		}

		// Look for module:app pattern
		if strings.Contains(arg, ":") {
			module := strings.Split(arg, ":")[0]
			return NewServiceMetadata(module, CommandLine), true
		}
	}
	return NewServiceMetadata("uvicorn", CommandLine), true
}

File: pkg/discovery/module/rust/src/service_name/python.rs
// Unless explicitly stated otherwise all files in this repository are licensed
// under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2025-present Datadog, Inc.

//! This package implements Python service name generation.

use std::iter::{Peekable, successors};
use std::os::unix::ffi::OsStrExt;
use std::path::{Path, PathBuf};

use normalize_path::NormalizePath;

use crate::fs::SubDirFs;
use crate::procfs::Cmdline;
use crate::service_name::{
    DetectionContext, ServiceNameMetadata, ServiceNameSource, gunicorn, uvicorn,
};

pub fn extract_name(cmdline: &Cmdline, ctx: &mut DetectionContext) -> Option<ServiceNameMetadata> {
    let mut args = cmdline.args().peekable();

    // Skip the first arg (python executable)
    args.next()?;

    // When Gunicorn is invoked via its wrapper script the command line ends up
    // looking like the example below, so redirect to the Gunicorn detector for
    // this case:
    //  /usr/bin/python3 /usr/bin/gunicorn foo:app()
    //
    // Another case where we want to redirect to the Gunicorn detector is when
    // gunicorn replaces its command line with something like the below. Because
    // of the [ready], we end up here first instead of going directly to the
    // Gunicorn detector.
    //  [ready] gunicorn: worker [airflow-webserver]
    match Path::new(args.peek()?).file_name()?.as_bytes() {
        b"gunicorn" | b"gunicorn:" => {
            args.next(); // Skip the gunicorn executable
            return Some(gunicorn::extract_name_from_args(args, &ctx.envs));
        }
        b"uvicorn" => {
            args.next(); // Skip the uvicorn executable
            return uvicorn::extract_name_from_args(args);
        }
        _ => {}
    };

    match parse_args(args)? {
        ArgType::Module(module) => Some(ServiceNameMetadata::new(
            module,
            ServiceNameSource::CommandLine,
        )),
        ArgType::FileName(path) => {
            // Normalize paths to handle .. and . components (e.g., /../example.py -> /example.py)
            let path = path.normalize();
            let path = ctx.resolve_working_dir_relative_path(&path).unwrap_or(path);

            let metadata = ctx.fs.metadata(&path).ok()?;
            let is_file = !metadata.is_dir();

            if is_file {
                let parent = path.parent();
                if parent.is_none_or(|p| p.as_os_str().is_empty() || p.as_os_str() == "/") {
                    // Path is a root level file, return the file name
                    return Some(ServiceNameMetadata::new(
                        find_nearest_top_level(&path, ctx.fs),
                        ServiceNameSource::CommandLine,
                    ));
                }
            }

            if let Some(name) = deduce_package_name(&path, is_file, ctx.fs) {
                return Some(ServiceNameMetadata::new(name, ServiceNameSource::Python));
            }

            // If deduce_package_name didn't find anything, fall back to find_nearest_top_level
            // For directories, check the directory itself
            // For files, check the parent directory
            let check_path = if is_file { path.parent()? } else { &path };

            let name = find_nearest_top_level(check_path, ctx.fs);
            if name != "." && name != "/" && name != "bin" && name != "sbin" {
                return Some(ServiceNameMetadata::new(
                    name,
                    ServiceNameSource::CommandLine,
                ));
            }

            Some(ServiceNameMetadata::new(
                find_nearest_top_level(Path::new(path.file_name()?), ctx.fs),
                ServiceNameSource::CommandLine,
            ))
        }
    }
}

#[derive(Debug, PartialEq)]
enum ArgType {
    Module(String),
    FileName(PathBuf),
}

fn parse_args<'a>(mut args: Peekable<impl Iterator<Item = &'a str>>) -> Option<ArgType> {
    let args = args.by_ref();
    'args_loop: while let Some(arg) = args.next() {
        if let Some(arg) = arg.strip_prefix("--") {
            if arg == "check-hash-based-pycs" {
                // Only long arg with an argument.  CPython doesn't allow
                // including the argument with an equals sign in the same arg.
                args.next();
            }

            continue;
        }

        if let Some(arg) = arg.strip_prefix("-") {
            for (i, c) in arg.char_indices() {
                let rest = arg.get(i + c.len_utf8()..)?;
                match c {
                    // Everything after -c is a command and it terminates
                    // the option parsing.
                    'c' => return None,
                    'm' => {
                        return if rest.is_empty() {
                            Some(ArgType::Module(args.next()?.to_string()))
                        } else {
                            Some(ArgType::Module(rest.to_string()))
                        };
                    }
                    'X' | 'W' => {
                        // Takes an argument, either attached here or in the next arg.
                        if rest.is_empty() {
                            args.next();
                            continue 'args_loop;
                        } else {
                            break;
                        }
                    }
                    _ => continue,
                }
            }
        } else {
            return Some(ArgType::FileName(PathBuf::from(arg)));
        }
    }

    None
}

/// Walks up the directory tree collecting directory names while `__init__.py` exists.
/// Returns the package name (directories joined with '.') if a package was found.
/// If `is_file` is true and directories were traversed, appends the filename (without extension) to the package.
fn deduce_package_name(path: &Path, is_file: bool, fs: &SubDirFs) -> Option<String> {
    let mut traversed = Vec::new();

    // Start from the directory containing the file, or the directory itself
    let start_dir = if is_file { path.parent()? } else { path };

    let parents = successors(Some(start_dir), |path| {
        let parent = path.parent()?;
        if parent.as_os_str().is_empty() {
            None
        } else {
            Some(parent)
        }
    });

    for parent in parents {
        if !fs.exists(parent.join("__init__.py")) {
            break;
        }

        // Prepend the directory name to traversed list
        if let Some(dir_name) = parent.file_name().and_then(|n| n.to_str()) {
            traversed.insert(0, dir_name);
        }
    }

    // Return None if no package directories were found
    if traversed.is_empty() {
        return None;
    }

    // If path is a file, append the filename without extension
    if is_file && let Some(mod_name) = path.file_stem() {
        traversed.push(mod_name.to_str()?);
    }

    Some(traversed.join("."))
}

/// Finds the nearest top-level directory that contains Python files by walking up from the given path.
/// Returns the basename of that directory without the extension.
/// If the path is a file, returns the filename without the extension.
fn find_nearest_top_level(path: &Path, fs: &SubDirFs) -> String {
    // Walk up the directory tree, collecting directories that contain .py files.
    // Stop when we reach an empty path (from relative paths) or when there are no .py files.
    // If path is a file, take_while filters it out (has_py_files returns false for files),
    // so we fall back to the original path via unwrap_or.
    let last_with_py = successors(Some(path), |path| {
        let parent = path.parent()?;
        // Stop at empty paths (e.g., parent of "file.py" is "")
        if parent.as_os_str().is_empty() {
            None
        } else {
            Some(parent)
        }
    })
    .take_while(|current| has_py_files(current, fs))
    .last()
    .unwrap_or(path);

    // Get the file stem (basename without extension)
    last_with_py
        .file_stem()
        .and_then(|s| s.to_str())
        .unwrap_or("")
        .to_string()
}

/// Checks if a directory contains any .py files
/// Returns false if path is not a directory.
fn has_py_files(path: &Path, fs: &SubDirFs) -> bool {
    fs.read_dir(path)
        .ok()
        .map(|entries| {
            entries.filter_map(Result::ok).any(|entry| {
                entry
                    .file_name()
                    .to_str()
                    .is_some_and(|name| name.ends_with(".py"))
            })
        })
        .unwrap_or(false)
}

#[cfg(test)]
#[allow(clippy::expect_used, clippy::unwrap_used)]
mod tests {
    use super::*;
    use crate::cmdline;
    use crate::test_utils::TestDataFs;
    use std::collections::HashMap;
    use std::path::PathBuf;

    fn test_ctx() -> (HashMap<String, String>, TestDataFs) {
        let fs = TestDataFs::new("python");
        let mut envs = HashMap::new();
        // Set PWD to root so that resolve_working_dir_relative_path can resolve absolute paths
        envs.insert("PWD".to_string(), "/".to_string());
        (envs, fs)
    }

    #[test]
    fn test_3rd_level_module_path() {
        let (envs, fs) = test_ctx();
        let mut ctx = DetectionContext::new(0, envs, fs.as_ref());
        let cmdline = cmdline!["python", "modules/m1/first/nice/package"];

        let result = extract_name(&cmdline, &mut ctx);
        assert!(result.is_some());
        let metadata = result.unwrap();
        assert_eq!(metadata.name, "m1.first.nice.package");
    }

    #[test]
    fn test_2nd_level_module_path() {
        let (envs, fs) = test_ctx();
        let mut ctx = DetectionContext::new(0, envs, fs.as_ref());
        let cmdline = cmdline!["python", "modules/m1/first/nice"];

        let result = extract_name(&cmdline, &mut ctx);
        assert!(result.is_some());
        let metadata = result.unwrap();
        assert_eq!(metadata.name, "m1.first.nice");
    }

    #[test]
    fn test_2nd_level_explicit_script_inside_module() {
        let (envs, fs) = test_ctx();
        let mut ctx = DetectionContext::new(0, envs, fs.as_ref());
        let cmdline = cmdline!["python", "modules/m1/first/nice/something.py"];

        let result = extract_name(&cmdline, &mut ctx);
        assert!(result.is_some());
        let metadata = result.unwrap();
        assert_eq!(metadata.name, "m1.first.nice.something");
    }

    #[test]
    fn test_1st_level_module_path() {
        let (envs, fs) = test_ctx();
        let mut ctx = DetectionContext::new(0, envs, fs.as_ref());
        let cmdline = cmdline!["python", "modules/m1/first"];

        let result = extract_name(&cmdline, &mut ctx);
        assert!(result.is_some());
        let metadata = result.unwrap();
        assert_eq!(metadata.name, "m1.first");
    }

    #[test]
    fn test_empty_module() {
        let (envs, fs) = test_ctx();
        let mut ctx = DetectionContext::new(0, envs, fs.as_ref());
        let cmdline = cmdline!["python", "modules/m2"];

        let result = extract_name(&cmdline, &mut ctx);
        assert!(result.is_some());
        let metadata = result.unwrap();
        assert_eq!(metadata.name, "m2");
    }

    #[test]
    fn test_main_in_a_dir() {
        let (envs, fs) = test_ctx();
        let mut ctx = DetectionContext::new(0, envs, fs.as_ref());
        let cmdline = cmdline!["python", "apps/app1"];

        let result = extract_name(&cmdline, &mut ctx);
        assert!(result.is_some());
        let metadata = result.unwrap();
        assert_eq!(metadata.name, "app1");
    }

    #[test]
    fn test_script_in_inner_dir() {
        let (envs, fs) = test_ctx();
        let mut ctx = DetectionContext::new(0, envs, fs.as_ref());
        let cmdline = cmdline!["python", "apps/app2/cmd/run.py"];

        let result = extract_name(&cmdline, &mut ctx);
        assert!(result.is_some());
        let metadata = result.unwrap();
        assert_eq!(metadata.name, "app2");
    }

    #[test]
    fn test_script_in_top_level_dir() {
        let (envs, fs) = test_ctx();
        let mut ctx = DetectionContext::new(0, envs, fs.as_ref());
        let cmdline = cmdline!["python", "apps/app2/setup.py"];

        let result = extract_name(&cmdline, &mut ctx);
        assert!(result.is_some());
        let metadata = result.unwrap();
        assert_eq!(metadata.name, "app2");
    }

    #[test]
    fn test_top_level_script() {
        let (envs, fs) = test_ctx();
        let mut ctx = DetectionContext::new(0, envs, fs.as_ref());
        let cmdline = cmdline!["python", "example.py"];

        let result = extract_name(&cmdline, &mut ctx);
        assert!(result.is_some());
        let metadata = result.unwrap();
        assert_eq!(metadata.name, "example");
    }

    #[test]
    fn test_root_level_script() {
        let (envs, fs) = test_ctx();
        let mut ctx = DetectionContext::new(0, envs, fs.as_ref());
        let cmdline = cmdline!["python", "/example.py"];

        let result = extract_name(&cmdline, &mut ctx);
        assert!(result.is_some());
        let metadata = result.unwrap();
        assert_eq!(metadata.name, "example");
    }

    #[test]
    fn test_root_level_script_with_dotdot() {
        let (envs, fs) = test_ctx();
        let mut ctx = DetectionContext::new(0, envs, fs.as_ref());
        let cmdline = cmdline!["python", "/../example.py"];

        let result = extract_name(&cmdline, &mut ctx);
        assert!(result.is_some());
        let metadata = result.unwrap();
        assert_eq!(metadata.name, "example");
    }

    #[test]
    fn test_script_in_bin() {
        let (envs, fs) = test_ctx();
        let mut ctx = DetectionContext::new(0, envs, fs.as_ref());
        let cmdline = cmdline!["python", "/usr/bin/pytest"];

        let result = extract_name(&cmdline, &mut ctx);
        assert!(result.is_some());
        let metadata = result.unwrap();
        assert_eq!(metadata.name, "pytest");
    }

    #[test]
    fn test_script_in_bin_with_u_flag() {
        let (envs, fs) = test_ctx();
        let mut ctx = DetectionContext::new(0, envs, fs.as_ref());
        let cmdline = cmdline!["python3", "-u", "bin/WALinuxAgent.egg"];

        let result = extract_name(&cmdline, &mut ctx);
        assert!(result.is_some());
        let metadata = result.unwrap();
        assert_eq!(metadata.name, "WALinuxAgent");
    }

    #[test]
    fn test_parse_python_args() {
        let tests: Vec<(&str, Cmdline, Option<ArgType>)> = vec![
            ("empty args", cmdline![], None),
            ("dash only", cmdline!["-"], None),
            ("-B flag only", cmdline!["-B"], None),
            ("-X flag only", cmdline!["-X"], None),
            (
                "script.py",
                cmdline!["script.py"],
                Some(ArgType::FileName(PathBuf::from("script.py"))),
            ),
            (
                "-u script.py",
                cmdline!["-u", "script.py"],
                Some(ArgType::FileName(PathBuf::from("script.py"))),
            ),
            (
                "-B script.py",
                cmdline!["-B", "script.py"],
                Some(ArgType::FileName(PathBuf::from("script.py"))),
            ),
            (
                "-E script.py",
                cmdline!["-E", "script.py"],
                Some(ArgType::FileName(PathBuf::from("script.py"))),
            ),
            (
                "-OO script.py",
                cmdline!["-OO", "script.py"],
                Some(ArgType::FileName(PathBuf::from("script.py"))),
            ),
            (
                "script.py -OO",
                cmdline!["script.py", "-OO"],
                Some(ArgType::FileName(PathBuf::from("script.py"))),
            ),
            (
                "-u -B -E script.py",
                cmdline!["-u", "-B", "-E", "script.py"],
                Some(ArgType::FileName(PathBuf::from("script.py"))),
            ),
            (
                "-W default app.py",
                cmdline!["-W", "default", "app.py"],
                Some(ArgType::FileName(PathBuf::from("app.py"))),
            ),
            (
                "-Wdefault app.py",
                cmdline!["-Wdefault", "app.py"],
                Some(ArgType::FileName(PathBuf::from("app.py"))),
            ),
            (
                "-c print('hello') app.py",
                cmdline!["-c", "print('hello')", "app.py"],
                None,
            ),
            (
                "-cprint('hello') app.py",
                cmdline!["-cprint('hello')", "app.py"],
                None,
            ),
            (
                "-m module app.py",
                cmdline!["-m", "module", "app.py"],
                Some(ArgType::Module("module".to_string())),
            ),
            (
                "-mmodule app.py",
                cmdline!["-mmodule", "app.py"],
                Some(ArgType::Module("module".to_string())),
            ),
            (
                "-BEmmodule -u script.py",
                cmdline!["-BEmmodule", "-u", "script.py"],
                Some(ArgType::Module("module".to_string())),
            ),
            (
                "--check-hash-based-pycs always app.py",
                cmdline!["--check-hash-based-pycs", "always", "app.py"],
                Some(ArgType::FileName(PathBuf::from("app.py"))),
            ),
            (
                "--foo script.py",
                cmdline!["--foo", "script.py"],
                Some(ArgType::FileName(PathBuf::from("script.py"))),
            ),
            (
                "-BE script.py",
                cmdline!["-BE", "script.py"],
                Some(ArgType::FileName(PathBuf::from("script.py"))),
            ),
            (
                "-BEs -u script.py",
                cmdline!["-BEs", "-u", "script.py"],
                Some(ArgType::FileName(PathBuf::from("script.py"))),
            ),
            (
                "-BW ignore script.py",
                cmdline!["-BW", "ignore", "script.py"],
                Some(ArgType::FileName(PathBuf::from("script.py"))),
            ),
            (
                "-X dev script.py",
                cmdline!["-X", "dev", "script.py"],
                Some(ArgType::FileName(PathBuf::from("script.py"))),
            ),
            (
                "-Xdev script.py",
                cmdline!["-Xdev", "script.py"],
                Some(ArgType::FileName(PathBuf::from("script.py"))),
            ),
            (
                "-W error::DeprecationWarning script.py",
                cmdline!["-W", "error::DeprecationWarning", "script.py"],
                Some(ArgType::FileName(PathBuf::from("script.py"))),
            ),
            (
                "-Werror::DeprecationWarning script.py",
                cmdline!["-Werror::DeprecationWarning", "script.py"],
                Some(ArgType::FileName(PathBuf::from("script.py"))),
            ),
        ];

        for (name, cmdline, expected) in tests {
            let result = parse_args(cmdline.args().peekable());
            assert_eq!(result, expected, "Test case '{}' failed", name);
        }
    }
}
